controllers/cluster_controller.go

```diff
@@ -108,76 +110,109 @@ func (r *ClusterReconciler) reconcile(ctx context.Context, cluster *capi.Cluster
-							ListOptions: client.ListOptions{
-								Namespace:     cluster.GetNamespace(),
-								LabelSelector: selector,
-							},
-							DeleteOptions: client.DeleteOptions{
-								PropagationPolicy: &propagationPolicy,
-							},
-						})
-						if err != nil {
-							log.Error(err, fmt.Sprintf("unable to delete Apps for cluster %s/%s", cluster.Namespace, cluster.Name))
-							ErrorsTotal.WithLabelValues(cluster.Name, cluster.Namespace).Inc()
-							return requeue(), nil
-						}
+			if r.CApiProvider == "" {
+				if !hasChartAnnotations(cluster) {
+					err := r.Client.Delete(ctx, cluster, client.PropagationPolicy(propagationPolicy))
+					if err != nil {
+						log.Error(err, fmt.Sprintf("unable to delete cluster %s/%s", cluster.Namespace, cluster.Name))
+						ErrorsTotal.WithLabelValues(cluster.Name, cluster.Namespace).Inc()
+						return requeue(), nil
 					}
+					log.Info(fmt.Sprintf("Cluster %s/%s was deleted", cluster.Namespace, cluster.Name))
+					SuccessTotal.WithLabelValues(cluster.Name, cluster.Namespace).Inc()
+				} else {
+					log.Info("CAPI cluster detected, ignoring")
+				}
+			} else {
+				if hasChartAnnotations(cluster) {
+					log.Info(fmt.Sprintf("Provider = %s", r.CApiProvider))
+
+					app := &gsapplication.App{}
+					err := r.Client.Get(ctx, getClusterAppNamespacedName(cluster), app)
+					if err != nil && !apierrors.IsNotFound(err) {
+						return ctrl.Result{}, err
+					} else if err == nil {
+						// ignore GitOps-managed resources, ensure MC itself doesn't commit suicide
+						if _, ok := app.Labels[fluxLabel]; ok {
+							IgnoredTotal.WithLabelValues(cluster.Name, cluster.Namespace).Inc()
+							log.Info(fmt.Sprintf("Found label %s in App CR. Cluster %s/%s will be ignored for deletion", fluxLabel, cluster.Namespace, cluster.Name))
+							return ctrl.Result{}, nil
+						}
 
-					// Delete cluster ConfigMaps
-					{
-						err := r.Client.DeleteAllOf(ctx, &corev1.ConfigMap{}, &client.DeleteAllOfOptions{
-							ListOptions: client.ListOptions{
-								Namespace:     cluster.GetNamespace(),
-								LabelSelector: selector,
-							},
-							DeleteOptions: client.DeleteOptions{
-								PropagationPolicy: &propagationPolicy,
-							},
-						})
-						if err != nil {
-							log.Error(err, fmt.Sprintf("unable to delete ConfigMaps for cluster %s/%s", cluster.Namespace, cluster.Name))
-							ErrorsTotal.WithLabelValues(cluster.Name, cluster.Namespace).Inc()
-							return requeue(), nil
+						// delete Cluster App CR
+						clusterAppSelector := labels.NewSelector()
+						clusterAppNameReq, _ := labels.NewRequirement(label.AppName, selection.In, []string{fmt.Sprintf("cluster-%s", r.CApiProvider)})
+						fluxLabelReq, _ := labels.NewRequirement(fluxLabel, selection.NotEquals, []string{"flux"})
+						clusterAppSelector = clusterAppSelector.Add(*clusterAppNameReq, *fluxLabelReq)
+						{
+							err := r.Client.DeleteAllOf(ctx, &gsapplication.App{}, &client.DeleteAllOfOptions{
```

Comment:
I may be missing something. Wouldn't this delete all `cluster-aws` apps on that namespace not managed by flux? If I understand correctly, we only want to delete the cluster that we are reconciling in this loop.